### PR TITLE
WIP: Discussion: multi-environment testing

### PR DESCRIPTION
**Pull request by dvoytenko**
Some key points:
- Note that `describe.js` is extremely dirty and hacky right now - subject to rewrite in a real PR.
- Test `describeEnv` describes in some detail what's tested. For instance, it includes a set of extensions that need to be installed for the test.
- `describeEnv` automatically runs all child tests against all environments. Currently a set of environments include: 
  - Standalone from direct origin and from CDN
  - Iframe from direct origin and from CDN
  - Webview from direct origin and from CDN
  - Shadow DOM from direct origin
  - Friendly iframe from direct origin and from CDN
- All extension installation code is moved into `AMP.extension()` function as can be seen in `amp-app-banner.js`. This allows installing only requested extensions in the test environment and doing so hermetically. The `AMP.extension()` is inlined in the PROD binary. It's functionally equivalent to our `AMP.push()`.
- The `createIframePattern` is essentially executed automatically and blocks `beforeEach` to remove unnecessary async calls in all `it` tests. When test starts, its environment is completely ready to run tests against.
- Tests always (or by default) start with runtime off. This means that build and layout have to be called on a tested element manually in test. We will provide `buildForTesting` and `layoutForTesting` callbacks in the `Resources` class for this. This would be better than waiting for random times in tests for this callbacks to be called or calling these methods twice as we do now.
- Mutli-env tests can be used for most of parts of AMP, including elements and services. However, where convenient, we will also provide `describeFakeWindow` to test against a fake window.

The main question I have is if that might be an overkill to potentially be testing most of AMP against 
10+ environments thus multiplying number of tests times 10. But I think it's worth to expand test coverage to scenarios that we'd otherwise not think about. And for local development, we can always add switches to only test against one environment by default to reduce local development time.

### CODE REVIEW

File: extensions/amp-app-banner/0.1/amp-app-banner.js [hunk 1]
```diff
@@ -435,4 +435,6 @@ function updateViewportPadding(state) {
 }
 
 
-AMP.registerElement('amp-app-banner', AmpAppBanner, CSS);
+AMP.extension('amp-app-banner', function(AMP) {
+  AMP.registerElement('amp-app-banner', AmpAppBanner, CSS);
```
Comment (cramforce) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Not a fan of the redundancy of `amp-app-banner`. Can be fixed, of course.

File: extensions/amp-app-banner/0.1/test/test-amp-app-banner.js [hunk 2]
```diff
@@ -70,61 +71,6 @@ describe('amp-app-banner', () => {
     }
   }
 
-  function getTestFrame() {
```
Comment (cramforce) [inline extensions/amp-app-banner/0.1/test/test-amp-app-banner.js]: Nice

**Comment by cramforce** [comment]
Worth it for the boilerplate reduction alone.

File: extensions/amp-app-banner/0.1/amp-app-banner.js (hunk 1 above)
Comment (dvoytenko) [inline extensions/amp-app-banner/0.1/amp-app-banner.js]: Yeah. It's definitely annoying. This really means our current `AMP.push({n: 'amp-app-banner', f: ...})`

**Pull request closed without merging**